Implementing Superbuild Structure for CMaize Project

eminokic opened the pull request.

Commit:
Implementing project structure.

Commit:
Adding a shared-library and connecting an interface to a class.

Commit:
Fixing add library error by refactoring from top level cmakelists.txt to source folder cmakelists.txt

Commit:
Adding functions for the cmakelists.txt file in the source folder to install builds properly, next step should be to somehow implement the cmakelists in a way that is called by the top level cmakelists.txt file and exported/installed to the proper location.

ryanmrichard commented on LICENSE:
```diff
@@ -1,201 +0,0 @@
-                                 Apache License
```
Please put the license back

ryanmrichard commented on Shared-Library/hello-world.hpp:
```diff
@@ -0,0 +1,5 @@
+
+string say_hello();
+
+int main();
```
shared libraries don't have main functions, also there's no point in forward delcaring the main function since it defines the API to an executable and not a library.

ryanmrichard reviewed (changes requested):
do not check in the build directory. I'd recommend adding your build directory names it to the `.gitignore` file so you don't accidentally check them in

Commit:
Adding license back to cmakesuperbuild

Commit:
Reorganizing project structure, adding license back, and refactoring lower-level cmake lists.
Note: Error occurring at high level cmakelists.txt since no executable is found.

Commit:
Adding Boost Library to CMaize Superbuild. Boost is helpful as an external project to search through file systems more efficiently.

ryanmrichard commented on CMakeLists.txt:
```diff
@@ -0,0 +1,37 @@
+
+project(CMaizeSuperbuild LANGUAGES CXX)
+
+list(APPEND BOOST_COMPONENTS_REQUIRED filesystem system)
```
For now I recommend having no dependencies in this project. Once your superbuild works for your hello world library then you can make the library depend on CMaizeExamples/CMakePublic. I wouldn't add any other dependencies to the project beyond that.

ryanmrichard commented on external/upstream/boost/CMakeLists.txt:
```diff
@@ -0,0 +1,104 @@
+find_package(Boost ${Boost_MINIMUM_REQUIRED} QUIET COMPONENTS "${BOOST_COMPONENTS_REQUIRED}")
```
I could be wrong, but I'm guessing you copy/pasted this from somewhere. If that's the case you need to attribute the source. That said as my previous comment suggested you don't want to use Boost in this project.

ryanmrichard commented on src/hello-world.cpp:
```diff
@@ -0,0 +1,10 @@
+
+std::string say_hello() { return std::string("Hello, CMake superbuild!"); }
+
+int main() {
```
So two things. First, in C++ it's customary to put the definition of the main function in a file `main.cpp`. Second your CMakeLists.txt isn't building an executable (it's building a library) so there shouldn't be a main function (all hell's going to break loose when someone tries to link to this library because there can't be two main functions). We ultimately need a library. I'd avoid building an executable for now (we can build one later as a test, but it's going to add more complexity).

Commit:
Refactoring CMaize Superbuild Project and removing boost library dependency

Commit:
Refactoring superbuild project structure and the hello world class and interface.

ryanmrichard commented on CMakeLists.txt:
```diff
@@ -2,6 +2,10 @@ cmake_minimum_required(VERSION 3.5 FATAL_ERROR)
 
+# Required to enable unit testing
+include(CTest)
+enable_testing()
```
don't explicitly call `enable_testing()`. `include(CTest)` will call it if `BUILD_TESTING` is true

Commit:
Adding external project CMakePublic via github url

ryanmrichard commented on CMakeLists.txt:
```diff
@@ -2,6 +2,10 @@ cmake_minimum_required(VERSION 3.5 FATAL_ERROR)
+include(CTest)
+enable_testing()
+
 set(CMAKE_CXX_STANDARD 11)
```
You aren't using an C++11 features, so don't force it. Anyways the C++ standard should be set on a target not globally. Also you should never call `set()` for variables defined by CMake (such as you do on lines 9 to 11); doing so will override the user's options. Usually you want to use the `option` command (which will only set them if the user hasn't).

Commit:
Removing cmake enable testing

Commit:
Refactoring CMaizeSuperbuild model to configure CMakePublic dependency to CMaizeSuperbuild_core

Commit:
Adding fetch content

ryanmrichard commented on CMakeLists.txt:
```diff
@@ -0,0 +1,35 @@
+    INSTALL_COMMAND ""
+)
+
+ExternalProject_Add(CMakePublic
```
Let's not worry about CMakePublic until your library builds and installs. Also `ExternalProject_Add` has a configure step so don't add one.

ryanmrichard commented on src/CMakeLists.txt:
```diff
@@ -0,0 +1,30 @@
+
+install(TARGETS ${PROJECT_NAME}
+        EXPORT ${PROJECT_NAME}-targets
+        PUBLIC_HEADER DESTINATION ${CMAKE_INSTALL_PREFIX}/${CMAKE_INSTALL_INCLUDEDIR}
```
Pretty sure you need `include(GNUInstallDirs)` before you can use `${CMAKE_INSTALL_INCLUDEDIR}`

ryanmrichard commented on CMakeLists.txt:
```diff
@@ -0,0 +1,35 @@
+
+include(CTest)
+include(ExternalProject)
+include(FetchContent)
```
You don't need `FetchContent`. `ExternalProject_Add` can get souce code. As a disclaimer I think way down the road we'll actually want to use `FetchContent` to get the source code (it'll make it easier to build offline, and for the user to override the default dependency version), but let's not worry about that until we have the superbuild in place.

Commit:
Implementing a revised superbuild model along with a few refactoring items. Issue is currently building the cmake project as it is unable to recognize the tag for CMakePublic. Also noted that the external project add may not be needed later for the cmakepublic repository.

ryanmrichard commented on CMake/Depends/CMakePublic.cmake:
```diff
@@ -0,0 +1,19 @@
+if(NOT TARGET depends::CMakePublic)
```
Don't worry about CMakePublic for this PR. Let's make this PR about building and installing your hello world library via external project add. I'd recommend just deleting all of the CMakePublic stuff you added; it's not what we'll need down the road either.

ryanmrichard commented on CMake/Modules/SuperBuildDepends.cmake:
```diff
@@ -0,0 +1,62 @@
+# MIT License
```
For the purposes of this repo you shouldn't be using other people's work, so remove this file. This repo is going to be a textbook superbuild example with no frills.

ryanmrichard commented on CMakeLists.txt:
```diff
@@ -0,0 +1,57 @@
+include_directories(${CMAKE_CURRENT_SOURCE_DIR}/src)
+include_directories(${EXTERNAL_INSTALL_LOCATION}/Shared-Library)
+
+add_library(hello-world)
```
This needs to go back to how you had it. I'm not sure why you pulled it up into the driver `CMakeLists.txt`.

ryanmrichard commented on Shared-Library/hello-world.hpp:
```diff
@@ -0,0 +1,5 @@
+#pragma once
+#include <core/version.h>
```
Not sure what this header is, but you don't appear to be using it so I'd drop it. Also it should be `std::string say_hello();` I presume you were having a compiler error?

ryanmrichard reviewed (changes requested):
Overall your most recent changes seem to be a step backwards. For this PR we're trying to:

1. Build your hello world library as an external project
2. Test the hello world library that you built as an external project
3. Install the hello world library

Once we have that working I think we should merge the PR. Then we can focus on adding a dependency. If I were you I'd revert to the previously reviewed commit (looks like it was 90f66ebe87635f85cea251cf4963beded3bd2a41). Then focus on removing CMakePublic.

Commit:
Reverting back to basic CMake project with external project add for hello-world prior to pushing further changes.

Commit:
Refactoring project.

Commit:
Refactoring hello world to a basic dummy line function for implementation purposes.

ryanmrichard commented on CMakeLists.txt:
```diff
@@ -0,0 +1,19 @@
+
+include(ExternalProject)
+
+include_directories(${CMAKE_CURRENT_SOURCE_DIR}/src)
+include_directories(${EXTERNAL_INSTALL_LOCATION}/Shared-Library)
```
Why are these here?

eminokic replied:
They were from the older push I had, I kept them since I had the two folder in my project, but for the superbuild I will just use an external project add command to add the directories in from the repository.

ryanmrichard replied:
Gotcha. Yeah I don't think you want these here for the superbuild.

ryanmrichard commented on src/hello-world.cpp:
```diff
@@ -0,0 +1,8 @@
+#include <iostream>
+
+int HelloWorld (int argc, char** argv)
```
This isn't the API you declared in the header file.

eminokic replied:
I will go ahead and update that now.

Commit:
Refactoring hello world files

ryanmrichard commented on Shared-Library/hello-world.hpp:
```diff
@@ -1,4 +1,3 @@
-#pragma once
```
You still need the `#pragma once`

ryanmrichard commented on src/hello-world.cpp:
```diff
@@ -1,8 +1,5 @@
+#include <cstdlib>
 #include <iostream>
```
You're no longer using these headers.

Commit:
Removing unnecessary include directory functions.

Commit:
Adding pragma once header and removing unnecessary headers from hello-world.

Commit:
Adding a .gitignore for cmake generated from gitignore.io

Commit:
Refactoring sub folder project name.

ryanmrichard commented on src/hello-world.cpp:
```diff
@@ -0,0 +1,3 @@
+#include <string>
```
It's standard C/C++ practice to include the header file defining the API, i.e., `#include "hello-world.hpp"`, which also pulls in your public dependencies (`<string>`).

Commit:
Adding include hello-world header file

Commit:
Removing include string since it is now being drawn from the hello-world.hpp class

Commit:
Pushing a new header for the hello-world.cpp file.

Commit:
Adding GNUInstall Directory

ryanmrichard commented on src/hello-world.cpp:
```diff
@@ -0,0 +1,3 @@
+#include "../Shared-Library/hello-world.hpp"
```
This should be `#include "Shared-Library/hello-world.hpp"`. Admittedly It doesn't matter so much for source files since they aren't referenced in other phases of the build, but doing so will help make sure you get the include paths right.

eminokic replied:
I was getting an error building it and locating the hpp file unless it was specified as ../

ryanmrichard replied:
That's because your include path isn't setup correctly.

eminokic replied:
Would that just be a my computer thing, or is the project path not setup correctly?

ryanmrichard replied:
I'd argue it's a problem with your project structure, but you could equivalently argue it's a problem with your `src/CMakeLists.txt` setting the wrong include paths. Right now you have your includes in `Shared-Library`, but your `src/CMakeLists.txt` claims they're in `include`; the latter is the more canonical place to put them.

ryanmrichard commented on CMakeLists.txt:
```diff
@@ -0,0 +1,18 @@
+include(ExternalProject)
+
+include(GNUInstallDirs)
+
```
You need to define a staging directory

ryanmrichard replied:
Don't bother caching the value of the staging directory unless you have a reason to.

ryanmrichard commented on CMakeLists.txt:
```diff
@@ -0,0 +1,18 @@
+    CMAKE_CACHE_ARGS -DCMAKE_CXX_FLAGS:STRING="${CMAKE_CXX_FLAGS}"
+    BUILD_ALWAYS 1
+    INSTALL_COMMAND ""
+)
```
You need to install the library.

ryanmrichard commented on CMakeLists.txt:
```diff
@@ -0,0 +1,18 @@
+
+include(ExternalProject)
+
+include(GNUInstallDirs)
```
This needs to move down to the `src/CMakeLists.txt` file where you actually use it.

ryanmrichard commented on CMakeLists.txt:
```diff
@@ -0,0 +1,18 @@
+    CMAKE_ARGS -DCMAKE_CXX_COMPILER="${CMAKE_CXX_COMPILER}"
+               -DCMAKE_CXX_STANDARD="${CMAKE_CXX_STANDARD}"
+               -DCMAKE_CXX_EXTENSIONS="${CMAKE_CXX_EXTENSIONS}"
+               -DCMAKE_CXX_STANDARD_REQUIRED="${CMAKE_CXX_STANDARD_REQUIRED}"
```
The library should be setting the standard it needs.

eminokic replied:
So is this an unnecessary line of code as well similar to extensions?

ryanmrichard replied:
yes

ryanmrichard replied:
to clarify, passing `CMAKE_CXX_STANDARD` is okay for now as that's something user could have reasonably set, but I wouldn't expect them to pass `CMAKE_CXX_EXTENSIONS` or `CMAKE_CXX_STANDARD_REQUIRED` (nor should they)

ryanmrichard commented on CMakeLists.txt:
```diff
@@ -0,0 +1,18 @@
+    SOURCE_DIR "${CMAKE_CURRENT_LIST_DIR}/src"
+    CMAKE_ARGS -DCMAKE_CXX_COMPILER="${CMAKE_CXX_COMPILER}"
+               -DCMAKE_CXX_STANDARD="${CMAKE_CXX_STANDARD}"
+               -DCMAKE_CXX_EXTENSIONS="${CMAKE_CXX_EXTENSIONS}"
```
Do you really need to set this? I feel like CMake takes care of this automatically.

eminokic replied:
All the super build models I have seen use this structure, but I can look into this more.

ryanmrichard replied:
I'm referring to setting `CMAKE_CXX_EXTENSIONS`. I don't think you need to mess with it.

eminokic replied:
Sounds good, should me removed now.

ryanmrichard commented on src/CMakeLists.txt:
```diff
@@ -0,0 +1,28 @@
+cmake_minimum_required(VERSION 3.5.1)
```
You don't need this here. The top-level file should take care of this for the whole project.

ryanmrichard commented on src/CMakeLists.txt:
```diff
@@ -0,0 +1,28 @@
+
+project(CMaizeSuperbuild LANGUAGES CXX)
+
+add_library(${PROJECT_NAME} SHARED hello-world.cpp)
```
Don't hard code the library to a shared library.

ryanmrichard commented on src/CMakeLists.txt:
```diff
@@ -0,0 +1,28 @@
+cmake_minimum_required(VERSION 3.5.1)
+
+project(CMaizeSuperbuild LANGUAGES CXX)
```
This project should probably be named "HelloWorld" or something. The superbuild part is really in the top-level lists.

eminokic replied:
I was reading that when doing multiple level CMakeLists you should use the same project name and have it drawn as a prefix build. I can remove it still if you believe that is better for the super-build.

ryanmrichard replied:
I'd be curious to hear their reasoning for this (if they give it), but that sounds like a bad idea to me. The project name shows up in a couple places, notably in the logs. If you use the same name it's harder to distinguish between the driver build (the top-level) and the subbuilds. That's why I vote for different (and more descriptive) names.

eminokic replied:
You're right, I have changed the name till I collect more reasoning as to if it is beneficial or not to have one project name.

ryanmrichard reviewed (changes requested):
I'd recommend putting the source file in `src` and header file in `include` more akin to a normal GNU structure.

Commit:
Adding external project and GNUInstallDir to lower level CMakeLists

Commit:
Removing cmake minimum from lower level CMakeList

Commit:
Removing add library from lower level CMakeList

Commit:
Changing src folder project name

Commit:
Refactoring external project add function.

Commit:
Refactoring external project add function.

Commit:
Defining stage install directory. Would this be what you were referring to?

ryanmrichard commented on CMakeLists.txt:
```diff
@@ -0,0 +1,22 @@
+include(GNUInstallDirs)
+
+set(
+        CMAKE_INSTALL_PREFIX "${CMAKE_BINARY_DIR}/stage"
```
As a general rule of thumb you never want to call `set` with a variable that starts with `CMAKE`. Those variables are how users set options. If you call `set` with one of them you clobber the user's setting. In this case I recommend you call the staging directory something like `STAGING_DIRECTORY`.

eminokic replied:
Would you recommend it be 

CMAKE_INSTALL_PREFIX "${STAGING_DIRECTORY}/stage"
CMAKE_INSTALL_PREFIX "STAGING_DIRECTORY/stage"

or remove the CMAKE_INSTALL_PREFIX with STAGING_DIRECTORY

ryanmrichard replied:
`set(STAGING_DIRECTORY ${CMAKE_BINARY_DIR}/stage/)` then in `ExternalProject_Add` pass the staging directory as the install directory like  `CMAKE_ARGS -DCMAKE_INSTALL_PREFIX="${STAGING_DIRECTORY}"`

Commit:
Adding cmake_minimum_required function to src cmakelist.txt

Commit:
Setting a staging directory in top level cmakelists.txt

ryanmrichard commented on CMakeLists.txt:
```diff
@@ -0,0 +1,23 @@
+    SOURCE_DIR "${CMAKE_CURRENT_LIST_DIR}/src"
+    CMAKE_ARGS -DCMAKE_CXX_COMPILER="${CMAKE_CXX_COMPILER}"
+               -DCMAKE_CXX_STANDARD="${CMAKE_CXX_STANDARD}"
+               -DCMAKE_INSTALL_PREFIX:PATH="${CMAKE_INSTALL_PREFIX}"
```
You're setting `CMAKE_INSTALL_PREFIX` twice. This line needs to go away.

Commit:
Refactoring stage directory function.

Commit:
Refactoring external project add function

Commit:
Refactoring Project structure.

Commit:
Refactoring project structure.

ryanmrichard commented on hello_world/src/hello-world.cpp:
```diff
@@ -1,3 +1,3 @@
-#include "hello-world.hpp"
+#include "../include/hello-world.hpp"
```
Don't use includes relative the source file; includes should be relative to the include path. It should just be `#include "hello-world.hpp"`

Commit:
Refactoring #include header and cmakelist.txt

Commit:
Implementing Get Property function from the external project library.

Commit:
Including the staging directory

ryanmrichard commented on CMakeLists.txt:
```diff
@@ -0,0 +1,25 @@
+    INSTALL_COMMAND ""
+)
+
+ExternalProject_Get_Property(${PROJECT_NAME}_core install_dir)
```
I  presume you are trying to install? You already know where it is supposed to be installed to (`CMAKE_INSTALL_PREFIX`). For a superbuild, installing should just be a matter of copying the staging directory to the final destination.

eminokic replied:
Is this unnecessary then?

ryanmrichard replied:
yes

ryanmrichard commented on CMakeLists.txt:
```diff
@@ -0,0 +1,25 @@
+
+ExternalProject_Get_Property(${PROJECT_NAME}_core install_dir)
+
+include_directories(${CMAKE_BINARY_DIR}/stage)
```
Why is this here?

eminokic replied:
I thought in order to make use of the staging directory then it would be required to include it. I will look into refactoring that.

ryanmrichard replied:
make use of it how? This is telling CMake where to look for header files when it goes to build a C++ project, but you've already built all the C++ projects at this point.

Commit:
Refactoring top level cmakelists.txt

Commit:
Testing with the add executable function.

ryanmrichard commented on CMakeLists.txt:
```diff
@@ -18,4 +22,10 @@ ExternalProject_Add(
     INSTALL_COMMAND ""
+)
+
+add_executable(
```
If you're going to make an executable it's got to be done as an external project also. I wouldn't worry about making one until you've got the installation working. Once you get the installation working let's merge this PR, then we can worry about adding dependencies and/or an executable.

Commit:
Refactoring unnecessary code from top level cmakelist.txt

Commit:
Implementing install command , still working out path.

ryanmrichard commented on hello_world/src/CMakeLists.txt:
```diff
@@ -5,7 +5,7 @@ include(ExternalProject)
 
 set_target_properties(${PROJECT_NAME} PROPERTIES PUBLIC_HEADER
-        ${CMAKE_CURRENT_SOURCE_DIR}/include/hello-world.hpp
+        ../include/hello-world.hpp
```
You should always use absolute paths in CMake files. Also why did you move this file?

eminokic replied:
I am not sure why it moved let me check and fix that.

Commit:
Refactoring file location for lower level cmake list.

Commit:
Refactoring files and adding subdirectory for lower level cmake list.

Commit:
no message

ryanmrichard commented on CMakeLists.txt:
```diff
@@ -0,0 +1,27 @@
+               -DCMAKE_INSTALL_PREFIX="${STAGING_DIRECTORY}"
+    CMAKE_CACHE_ARGS -DCMAKE_CXX_FLAGS:STRING="${CMAKE_CXX_FLAGS}"
+    BUILD_ALWAYS 1
+        DESTDIR=${CMAKE_BINARY_DIR}/stage
```
This is just `STAGING_DIRECTORY`; also does `ExternalProject_Add` recognize `DESTDIR`?

ryanmrichard commented on hello_world/CMakeLists.txt:
```diff
@@ -0,0 +1,31 @@
+cmake_minimum_required(VERSION ${CMAKE_VERSION})
+project(HelloWorld LANGUAGES CXX)
+
+include(ExternalProject)
```
You don't need `ExternalProject` here

ryanmrichard commented on hello_world/CMakeLists.txt:
```diff
@@ -0,0 +1,31 @@
+include(ExternalProject)
+include(GNUInstallDirs)
+
+add_subdirectory(src)
```
don't `add_subdirectory` `src` and `include`, just define the library in this file.

Commit:
Refactoring lower level cmakelist

Commit:
Refactoring cmakelists.txt on both levels.

ryanmrichard commented:
I finished this off in PR #2. The next step is to add CMakePublic as a dependency to the super build.
